paasta_tools/autoscaling_lib.py

```diff
@@ -31,78 +42,157 @@ def get_autoscaling_method(function_name):
-
-
-def get_instances_from_zookeeper(service, instance):
-    with ZookeeperPool() as zookeeper_client:
-        (instances, _) = zookeeper_client.get('%s/instances' % compose_autoscaling_zookeeper_root(service, instance))
-        return int(instances)
-
-
 @register_autoscaling_method('bespoke')
 def bespoke_autoscaling_method(*args, **kwargs):
-    # do nothing, the service author has written their own scaling code
+    """
+    Autoscaling method for service authors that have written their own autoscaling method.
+    Allows Marathon to read instance counts from Zookeeper but doesn't attempt to scale the service automatically.
+    """
     return 0
 
 
 @register_autoscaling_method('default')
-def default_autoscaling_method(marathon_service_config):
-    # not implemented yet
-    return 0
+def default_autoscaling_method(marathon_service_config, marathon_client, delay=600, setpoint=0.8, **kwargs):
+    """
+    PID control of instance count using ram and cpu.
 
+    :param marathon_service_config: the MarathonServiceConfig to scale
+    :param marathon_client: a Marathon client to fetch task data
+    :param delay: the number of seconds to wait between PID updates
+    :param setpoint: the target utilization percentage
 
-def autoscale_marathon_instance(marathon_service_config):
-    if marathon_service_config.get_max_instances() is None:
-        return
+    :returns: the number of instances to scale up/down by
+    """
+    Kp = 0.2
+    Ki = 0.2 / delay
+    Kd = 0.05 * delay
+
+    def clamp_value(number):
+        """Limits the PID output to scaling up/down by 1.
+        Also used to limit the integral term which avoids windup lag."""
+        return min(max(number, -1), 1)
+
+    def get_current_time():
+        return int(datetime.now().strftime('%s'))
+
+    autoscaling_root = compose_autoscaling_zookeeper_root(
+        service=marathon_service_config.service,
+        instance=marathon_service_config.instance,
+    )
+    zk_iterm_path = '%s/iterm' % autoscaling_root
+    zk_last_error_path = '%s/last_error' % autoscaling_root
+    zk_last_time_path = '%s/last_time' % autoscaling_root
+    zk_last_cpu_data = '%s/cpu_data' % autoscaling_root
+
+    with ZookeeperPool() as zk:
+        try:
+            iterm, _ = zk.get(zk_iterm_path)
+            last_error, _ = zk.get(zk_last_error_path)
+            last_time, _ = zk.get(zk_last_time_path)
+            last_cpu_data, _ = zk.get(zk_last_cpu_data)
+            iterm = float(iterm)
+            last_error = float(last_error)
+            last_time = float(last_time)
+            last_cpu_data = (datum for datum in last_cpu_data.split(',') if datum)
+        except NoNodeError:
+            iterm = 0.0
+            last_error = 0.0
+            last_time = 0.0
+            last_cpu_data = []
+
+    if get_current_time() - last_time < delay:
+        return 0
+
+    job_id = marathon_service_config.format_marathon_app_dict()['id']
```

Comment:
@solarkennedy this also takes a long time for services that don't have constraints set - running `marathon_tools.get_constraints` without any constraints set calls `mesos_tools.get_mesos_slaves_grouped_by_attribute`, which can take up to 15 seconds.

I guess this can be avoided just by filtering by service name and instance name, and either not autoscaling tasks that have multiple versions up (bouncing) or by ignoring the version difference and treating them all as one task for the purposes of autoscaling.